Add Flask example and tests

greyli opened the pull request:
This is a basic example for Flask and some universal tests for all the examples. The app generally follows the design in #1, the only difference is that I remove the `age` field.

After it's merged, I will start to set up the repo (tox, pre-commit hook, CI, etc.).

jugmac00 commented on examples/flask/app.py:
```diff
@@ -0,0 +1,119 @@
+        return {'message': 'The name and category field is required.'}, 400
+    if len(data['name']) > 10:
+        return {'message': 'The length of the name must be between 0 and 10.'}, 400
+    if data['category'] not in ['cat', 'dog']:
+        return {'message': 'The category must be one of: dog, cat.'}, 400
```
- The categories are hardcoded.
- There is a duplication of the categories, one time in the list, one time in text.

Why not make a top-level list or set, so it could be overwritten?

jugmac00 commented on examples/flask/app.py:
```diff
@@ -0,0 +1,119 @@
+    data = request.json
+    if 'name' not in data or 'category' not in data:
+        return {'message': 'The name and category field is required.'}, 400
+    if len(data['name']) > 10:
```
Maybe this is a matter of personal preference, but why duplicate the field length all over the code, instead of relying on the database to throw an error, catch it, and prepare a nice message?

greyli replied:
I think it will be better and easier to read and maintain to put all the validation logic in one place instead of delaying it to the database level.

jugmac00 reviewed (approved):
First, thank you so much for putting so much effort into showcasing Flask's capabilities to provide web APIs.

A couple of remarks:
- README.md is missing
- requirements.txt is missing
- it would be nice to have some curl commands for copy pasting at the top level README to play with the API

Additionally, I have left some inline comments.

Btw, what is the reason to not support `PUT`?

Could you please explain the `indirect=True` parameter for `parametrize`? I had a look at `pytest`s docs, but I am still puzzled.
Disclaimer, I don't work a lot with REST API.

I used `curl` to query the API, and looks good from my side!
![](https://media.tenor.co/images/b64253f542857e925f636da6227eee35/tenor.gif)

greyli commented:
Thanks for the review, I just added the missing README.md and requirements.txt/in files, and moved error messages and category list to separate variables.

> it would be nice to have some curl commands for copy pasting at the top level README to play with the API

Sure, I will add it later.

> what is the reason to not support PUT?

The PUT for updating a resource will be duplicate most of the `update_pet` view, so I think only support PATCH is enough for an example application.

> Could you please explain the indirect=True parameter for parameterized?

It will pass the items in the `example` list to the `client` fixture (via `request.param`), so we can iterate all the example applications for each test.

jugmac00 commented on examples/flask/README.md:
````diff
@@ -0,0 +1,78 @@
+    pass
+```
+
+With `flask.views.MethodView`, you can also organize the APIs with class. Each
````
```suggestion
With `flask.views.MethodView`, you can also organize the APIs within a class. Each
```

greyli replied:
I think this place can be left as it is. The class-based views use `class` to organize the API instead of `function`.

jugmac00 commented on examples/flask/README.md:
````diff
@@ -0,0 +1,78 @@
+```
+
+With `flask.views.MethodView`, you can also organize the APIs with class. Each
+method of the class maps to the corresponding HTTP method when dispatching
````
```suggestion
method maps to the corresponding HTTP method when dispatching
```

`method` is a function in class scope - so I think there is no need to explicitly name that the methods are "of the class".

greyli replied:
I was thinking to keep consistent with this change in Flask docs: https://github.com/pallets/flask/pull/4175#pullrequestreview-693302166

jugmac00 commented on examples/flask/app.py:
```diff
@@ -0,0 +1,125 @@
+PET_CATEGORIES = ['cat', 'dog']
+PET_REQUIRED_FIELD_ERROR = 'The name and category field is required.'
+PET_NAME_LENGTH_ERROR = 'The length of the name must be between 0 and 10.'
+PET_CATEGORY_ERROR = 'The category must be one of: dog, cat.'
```
Maybe you want to generate the categories (behind the colon) from `PET_CATEGORIES`? But anyway, this is not production code, this is an example code :-)

greyli replied:
Some [tests](https://github.com/pallets-eco/flask-api-examples/pull/2/files#diff-0dcce8a12c3812a7f4ba2f71b39da7d4d39fabcd8a242f4e602255bdc41237d8R37) are fixed with the preset category of the fake data, they will fail if we use a randomly generated category. I moved the fake data into the `init_database` function to make it more clear and compact.

jugmac00 reviewed (approved):
Some more minor comments :+1:
